jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/DemandingMessageHandler.java

```diff
@@ -0,0 +1,208 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.util;
+
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.io.ByteBufferCallbackAccumulator;
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.Utf8Appendable;
+import org.eclipse.jetty.util.Utf8StringBuilder;
+import org.eclipse.jetty.websocket.core.CloseStatus;
+import org.eclipse.jetty.websocket.core.CoreSession;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.core.exception.BadPayloadException;
+import org.eclipse.jetty.websocket.core.exception.MessageTooLargeException;
+import org.eclipse.jetty.websocket.core.internal.AbstractMessageHandler;
+
+/**
+ * <p>A utility implementation of FrameHandler that de-fragments text frames and binary frames into a whole messages before
+ * calling {@link #onText(String, Callback)} or {@link #onBinary(ByteBuffer, Callback)}.</p>
+ * <p>This is a demanding frame handler so flow control is by not automatic, an implementation of {@link #onText(String, Callback)}
+ * and {@link #onBinary(ByteBuffer, Callback)} must call {@link CoreSession#demand(long)} to receive the next WebSocket frame.
+ * Because this is a demanding FrameHandler it can be more efficient for binary messages by avoiding copies
+ * for aggregation until the full message has been received.</p>
+ */
+public class DemandingMessageHandler extends AbstractMessageHandler
```

Comment:
`DemandingMessageHandler` means to me that this class demands, but apparently it's not the case.
I would rename the class to just `MessageHandler`.

Also, I don't particularly like that to demand the next _message_ you have to use an API that demands _frames_.
However, I understand this is a low-level utility class, so I guess users have to know about the low-levelness.

Please review the javadoc, contains typos and says it's a demanding FrameHandler, but it's not (or it is? but if it is, why all the methods need to call `demand(1)`?).

Reply:
It is named `DemandingMessageHandler` because `FrameHandler.isDemanding()` returns true. I have now renamed the other one to `AutoDemandingMessageHandler` to make it more clear.

I have also added a `demand()` method onto `DemandingMessageHandler`.

Not sure what these typos are you are referring to but I have added to the javadoc.

Reply:
@lachlan-roberts sorry don't understand. `isDemanding()==true` means that we have to explicitly demand, but `isDemanding()==false` means "auto-demanding"? Makes no sense.

Perhaps you should rename `FrameHandler.isDemanding()` to something that is clearer, so far it's just confusing to no end.

Reply:
@sbordet I don't think we can be just renaming public api methods like that and breaking anyone implementing `FrameHandler`.

It is currently explained what demanding means in the javadoc of `FrameHandler`. We could change the name of this for 12 if we have a better name for it. @gregw what do you think?